protocol/xmrmaker/event.go

```diff
@@ -0,0 +1,198 @@
+package xmrmaker
+
+import (
+	"fmt"
+	"reflect"
+
+	"github.com/athanorlabs/atomic-swap/common/types"
+	mcrypto "github.com/athanorlabs/atomic-swap/crypto/monero"
+	"github.com/athanorlabs/atomic-swap/net/message"
+)
+
+// getStatus returns the status corresponding to the next expected event.
+func getStatus(t Event) types.Status {
+	switch t.(type) {
+	case *EventETHLocked:
+		return types.KeysExchanged
+	case *EventContractReady:
+		return types.XMRLocked
+	default:
+		return types.UnknownStatus
+	}
+}
+
+// Event represents a swap state event.
+type Event interface{}
+
+// EventETHLocked is the first expected event. It represents ETH being locked
+// on-chain.
+type EventETHLocked struct {
+	message *message.NotifyETHLocked
+	errCh   chan error
+}
+
+func newEventETHLocked(msg *message.NotifyETHLocked) *EventETHLocked {
+	return &EventETHLocked{
+		message: msg,
+		errCh:   make(chan error),
+	}
+}
+
+// EventContractReady is the second expected event. It represents the contract being
+// ready for us to claim the ETH.
+type EventContractReady struct {
+	errCh chan error
+}
+
+func newEventContractReady() *EventContractReady {
+	return &EventContractReady{
+		errCh: make(chan error),
+	}
+}
+
+// EventETHRefunded is an optional event. It represents the ETH being refunded back
+// to the counterparty, and thus we also must refund.
+type EventETHRefunded struct {
+	sk    *mcrypto.PrivateSpendKey
+	errCh chan error
+}
+
+func newEventETHRefunded(sk *mcrypto.PrivateSpendKey) *EventETHRefunded {
+	return &EventETHRefunded{
+		sk:    sk,
+		errCh: make(chan error),
+	}
+}
+
+// EventExit is an optional event. It is sent when the protocol should be stopped,
+// for example if the remote peer closes their connection with us before sending all
+// required messages, or we decide to cancel the swap.
+type EventExit struct {
+	errCh chan error
+}
+
+func newEventExit() *EventExit {
+	return &EventExit{
+		errCh: make(chan error),
+	}
+}
+
+func (s *swapState) runHandleEvents() {
+	for {
+		select {
+		case <-s.ctx.Done():
+			return
+		case event := <-s.eventCh:
+			s.handleEvent(event)
+		}
+	}
+}
+
+func (s *swapState) handleEvent(event Event) {
+	// events are only used once, so their error channel can be closed after handling.
+	switch e := event.(type) {
+	case *EventETHLocked:
+		log.Infof("EventETHLocked")
+		defer close(e.errCh)
+
+		if reflect.TypeOf(s.nextExpectedEvent) != reflect.TypeOf(&EventETHLocked{}) {
+			e.errCh <- fmt.Errorf("nextExpectedEvent was %T, not %T", s.nextExpectedEvent, e)
+			return
+		}
+
+		err := s.handleEventETHLocked(e)
+		if err != nil {
+			e.errCh <- fmt.Errorf("failed to handle EventETHLocked: %w", err)
+			return
+		}
+
+		s.setNextExpectedEvent(&EventContractReady{})
```

Comment:
It is confusing that the `errCh` inside `EventContractReady` is not getting initialized. If all we need is the type for next expected event, maybe we should be storing an enumerated value or the type itself instead of an uninitialized object of the type?

Reply:
updated to have a `Type()` method